truffle/src/main/java/org/jruby/truffle/language/objects/ReadObjectFieldNode.java

```diff
@@ -58,6 +58,9 @@ public Object updateShapeAndRead(DynamicObject object) {
     @TruffleBoundary
     @Specialization(contains = { "readObjectFieldCached", "updateShapeAndRead" })
     protected Object readObjectFieldUncached(DynamicObject receiver) {
+        if(getProperty(receiver.getShape(), name) == null){
+            return defaultValue;
+        }
         return receiver.get(name, defaultValue);
```

Comment:
This line intended to be the same since http://lafo.ssw.uni-linz.ac.at/javadoc/truffle/latest/com/oracle/truffle/api/object/DynamicObject.html#get-java.lang.Object-java.lang.Object- should return `defaultValue` if the property does not exist.
However, this is a problem as we have a "removed property flag" and so we need to use `PropertyFlags.isDefined(property)`. I'll fix, thanks for the report and analysis!